service/src/main/java/uk/nhs/adaptors/gp2gp/ehr/mapper/AgentDirectoryMapper.java

```diff
@@ -0,0 +1,159 @@
+
+        return mappingTriplets.stream()
+            .map(agentData -> mapAgentPerson(agentData.getPractitioner(), agentData.getPractitionerRole(), agentData.getOrganization()))
+            .filter(StringUtils::isNotEmpty)
+            .collect(Collectors.toList());
+    }
+
+    private String buildAgentPerson(Observation observation, Bundle bundle) {
+        var practitioner = AgentDirectoryExtractor.extractObservationResource(observation, bundle, ResourceType.Practitioner);
+        var organization = AgentDirectoryExtractor.extractObservationResource(observation, bundle, ResourceType.Organization);
+        if (practitioner.isPresent() && organization.isPresent()) {
+            return mapAgentPerson((Practitioner) practitioner.get(),
+                null,
+                (Organization) organization.get());
+        } else if (practitioner.isPresent()) {
+            var agentDataList = AgentDirectoryExtractor.extractAgentData(bundle, List.of((Practitioner) practitioner.get()))
+                .stream()
+                .map(agentData -> mapAgentPerson(agentData.getPractitioner(), agentData.getPractitionerRole(), agentData.getOrganization()))
+                .filter(StringUtils::isNotEmpty)
+                .findFirst();
+            if (agentDataList.isPresent()) {
+                return agentDataList.get();
+            }
+        }
+        return StringUtils.EMPTY;
+    }
+
+    private String buildAgentPerson(ReferralRequest referralRequest, Bundle bundle) {
+        var practitioner = ResourceExtractor
+            .extractResourceByReference(bundle, referralRequest.getRequester().getAgent().getReferenceElement());
+        var organization = ResourceExtractor
+            .extractResourceByReference(bundle, referralRequest.getRequester().getOnBehalfOf().getReferenceElement());
+        if (practitioner.isPresent() && organization.isPresent()) {
+            return mapAgentPerson(practitioner.map(Practitioner.class::cast).get(),
+                null,
+                organization.map(Organization.class::cast).get());
+
+        } else if (practitioner.isPresent()) {
+            var agentDataList = AgentDirectoryExtractor.extractAgentData(bundle, List.of(practitioner.map(Practitioner.class::cast).get()))
+                .stream()
+                .map(agentData -> mapAgentPerson(agentData.getPractitioner(), agentData.getPractitionerRole(), agentData.getOrganization()))
+                .filter(StringUtils::isNotEmpty)
+                .findFirst();
+
+            if (agentDataList.isPresent()) {
+                return agentDataList.get();
+            }
+        }
+        return StringUtils.EMPTY;
+    }
+
+    private String mapAgentPerson(Practitioner practitioner, PractitionerRole practitionerRole, Organization organization) {
+        if (!organizationPractitionerHasBeenMapped(practitioner, organization)) {
+            addOrganizationPractitionerPairToMap(practitioner, organization);
```

Comment:
Can we retrieve the `mappedId` before calling `organizationPractitionerHasBeenMapped` and `addOrganizationPractitionerPairToMap`, rather than duplicating the mappedId logic across both methods?